File: array/1329_sort_the_matrix_diagonally/1329_sort_the_matrix_diagonally.py
from collections import defaultdict


class Solution(object):
    def diagonalSort(self, mat):
        diag = defaultdict(list)
        N, M = len(mat), len(mat[0])

        for i in range(N):
            for j in range(M):
                diag[i-j].append(mat[i][j])
        print diag
        for k in diag.keys():
            diag[k].sort(reverse=True)
        print diag
        for i in range(N):
            for j in range(M):
                mat[i][j] = diag[i-j].pop()

        return mat



mat = [[3,3,1,1],[2,2,1,2],[1,1,1,2]]

res = Solution().diagonalSort(mat)
print(res)

File: array/1329_sort_the_matrix_diagonally/1329_sort_the_matrix_diagonally.go
package main

import (
	"fmt"
	sort2 "sort"
)

func diagonalSort(mat [][]int) [][]int {
	var (
		i    int
		lenX = len(mat)
	)
	if lenX == 0 {
		return mat
	}

	var (
		lenY = len(mat[0])
	)

	for i = 0; i < lenX; i++ {
		sort(mat, lenX-1-i, 0, lenX, lenY)
	}
	for i = 1; i < lenY; i++ {
		sort(mat, 0, i, lenX, lenY)
	}
	return mat
}

func sort(mat [][]int, x, y, lenX, lenY int) {
	var (
		tmp        []int
		tmpX, tmpY = x, y
	)
	for tmpX < lenX && tmpY < lenY {
		tmp = append(tmp, mat[tmpX][tmpY])
		tmpX++
		tmpY++
	}
	tmpX = 0
	sort2.Ints(tmp)
	for x < lenX && y < lenY {
		mat[x][y] = tmp[tmpX]
		tmpX++
		x++
		y++
	}
}

func main() {
	mat := [][]int{{3, 3, 1, 1}, {2, 2, 1, 2}, {1, 1, 1, 2}}
	res := diagonalSort(mat)
	fmt.Println(res)
}
